message/validation/validation.go

```diff
@@ -0,0 +1,559 @@
+	Signers         []spectypes.OperatorID
+	Committee       []*spectypes.Operator
+}
+
+// Descriptor provides details about a message. It's used for logging and metrics.
+type Descriptor struct {
+	ValidatorPK    spectypes.ValidatorPK
+	Role           spectypes.BeaconRole
+	SSVMessageType spectypes.MsgType
+	Slot           phase0.Slot
+	Consensus      *ConsensusDescriptor
+}
+
+// Fields returns zap logging fields for the descriptor.
+func (d Descriptor) Fields() []zapcore.Field {
+	result := []zapcore.Field{
+		fields.Validator(d.ValidatorPK),
+		fields.Role(d.Role),
+		zap.String("ssv_message_type", ssvmessage.MsgTypeToString(d.SSVMessageType)),
+		fields.Slot(d.Slot),
+	}
+
+	if d.Consensus != nil {
+		var committee []spectypes.OperatorID
+		for _, o := range d.Consensus.Committee {
+			committee = append(committee, o.OperatorID)
+		}
+
+		result = append(result,
+			fields.Round(d.Consensus.Round),
+			zap.String("qbft_message_type", ssvmessage.QBFTMsgTypeToString(d.Consensus.QBFTMessageType)),
+			zap.Uint64s("signers", d.Consensus.Signers),
+			zap.Uint64s("committee", committee),
+		)
+	}
+
+	return result
+}
+
+// String provides a string representation of the descriptor. It may be useful for logging.
+func (d Descriptor) String() string {
+	sb := strings.Builder{}
+	sb.WriteString(fmt.Sprintf("validator PK: %v, role: %v, ssv message type: %v, slot: %v",
+		hex.EncodeToString(d.ValidatorPK),
+		d.Role.String(),
+		ssvmessage.MsgTypeToString(d.SSVMessageType),
+		d.Slot,
+	))
+
+	if d.Consensus != nil {
+		var committee []spectypes.OperatorID
+		for _, o := range d.Consensus.Committee {
+			committee = append(committee, o.OperatorID)
+		}
+
+		sb.WriteString(fmt.Sprintf(", round: %v, qbft message type: %v, signers: %v, committee: %v",
+			d.Consensus.Round,
+			ssvmessage.QBFTMsgTypeToString(d.Consensus.QBFTMessageType),
+			d.Consensus.Signers,
+			committee,
+		))
+	}
+
+	return sb.String()
+}
+
+// ValidatorForTopic returns a validation function for the given topic.
+// This function can be used to validate messages within the libp2p pubsub framework.
+func (mv *messageValidator) ValidatorForTopic(_ string) func(ctx context.Context, p peer.ID, pmsg *pubsub.Message) pubsub.ValidationResult {
+	return mv.ValidatePubsubMessage
+}
+
+// ValidatePubsubMessage validates the given pubsub message.
+// Depending on the outcome, it will return one of the pubsub validation results (Accept, Ignore, or Reject).
+func (mv *messageValidator) ValidatePubsubMessage(_ context.Context, _ peer.ID, pmsg *pubsub.Message) pubsub.ValidationResult {
+	start := time.Now()
+	var validationDurationLabels []string // TODO: implement
+
+	defer func() {
+		sinceStart := time.Since(start)
+		mv.metrics.MessageValidationDuration(sinceStart, validationDurationLabels...)
+	}()
+
+	decodedMessage, descriptor, err := mv.validateP2PMessage(pmsg, time.Now())
+	round := specqbft.Round(0)
+	if descriptor.Consensus != nil {
+		round = descriptor.Consensus.Round
+	}
```

Comment:
it's a bit weird that you are not checking the err first, I assume descriptor can have value even if the err is not nil, but this is also not a very common usage.

Reply:
I agree it looks sketchy but the purpose of the descriptor is for logging, so `round` is used in error handling